controllers/cassandradatacenter_controller_test.go

```diff
@@ -0,0 +1,69 @@
+package controllers
+
+import (
+	"context"
+	cassdcapi "github.com/datastax/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	CassandraDatacenterName = "dc1"
+	ControllerTestNamespace = "dc-test"
+)
+
+var _ = Describe("Verify functionality of CassandraDatacenterReconciler", func() {
+	Specify("..", func() {
+		testNamespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: ControllerTestNamespace,
+			},
+		}
+		Expect(k8sClient.Create(context.Background(), testNamespace)).Should(Succeed())
+		reaper := createReaper(ControllerTestNamespace)
+		Expect(k8sClient.Create(context.Background(), reaper)).Should(Succeed())
+
+		testDc := &cassdcapi.CassandraDatacenter{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      CassandraDatacenterName,
+				Namespace: ControllerTestNamespace,
+				Annotations: map[string]string{
+					"reaper.cassandra-reaper.io/instance": "notHere",
+				},
+			},
+			Spec: cassdcapi.CassandraDatacenterSpec{
+				ClusterName:   "test-dc",
+				ServerType:    "cassandra",
+				ServerVersion: "3.11.7",
+				Size:          3,
+			},
+			Status: cassdcapi.CassandraDatacenterStatus{
+				CassandraOperatorProgress: cassdcapi.ProgressReady,
+				Conditions: []cassdcapi.DatacenterCondition{
+					{
+						Status: corev1.ConditionTrue,
+						Type:   cassdcapi.DatacenterReady,
+					},
+				},
+			},
+		}
+		Expect(k8sClient.Create(context.Background(), testDc)).Should(Succeed())
+
+		By("update annotation to target existing reaper instance")
+		testDcPatch := client.MergeFrom(testDc.DeepCopy())
+		testDc.Annotations = map[string]string{
+			"reaper.cassandra-reaper.io/instance": reaper.Name,
+		}
+		Expect(k8sClient.Patch(context.Background(), testDc, testDcPatch)).Should(Succeed())
+
+		By("make Reaper ready")
+		reaper.Status.Ready = true
+		reaper.Status.Clusters = append(reaper.Status.Clusters, "test-dc")
+		k8sClient.Status().Update(context.Background(), reaper)
+		verifyReaperReady(types.NamespacedName{Namespace: ControllerTestNamespace, Name: ReaperName})
```

Comment:
I am not sure `verifyReaperReady` call here. Here are some scenarios I think we want to test for and verify (if possible):

* The CassandraDatacenter does **not** have the `reaper.cassandra-reaper.io/instance` annotation
    * Verify that the Reaper instance is not modified
* The CassandraDatacenter has the `reaper.cassandra-reaper.io/instance` annotation and points to a non-existent Reaper instance
    * Verify that Reaper instance is not modified
* The CassandraDatacenter has the `reaper.cassandra-reaper.io/instance` annotation and the Reaper instance is not ready
    * Verify that the Reaper instance is not modified
* The CassandraDatacenter has the `reaper.cassandra-reaper.io/instance` annotation and the Reaper instance is  ready
    * Adding cluster to Reaper succeeds
        * Verify that the cluster is added in Reaper status
    * Adding cluster to Reaper fails
        * Verify that the cluster is not added to Reaper status

Implementing these tests is more challenging because we have the Reaper REST API calls. We could deal with that in one of two ways - stand up a Reaper instance or mock/stub `restClient`. I would go with the latter since we have coverage of `restClient` in the e2e tests.

We call `reapergo.NewReaperClient()` directly in the `Reconcile` function. If want to provide a different implementation, we need a small layer of indirection. We need a function or interface that is responsible for creating the REST client which we swap out during the integration test. 

We would get a good bit of additional test coverage at the cost of a bit of indirection.

Thoughts?

Reply:
Yeah, mocking or making a fake http server (which responds what the client wants) is the only option that can be used as the testing stops otherwise when the reaperClient is called.

Reply:
As for why verifyReaperReady, it's required for the controller to pass this section:

```go
		if !reaper.Status.Ready {
			r.Log.Info("waiting for reaper to become ready", "reaper", reaperKey)
			return ctrl.Result{RequeueAfter: shortDelay}, nil
		}
```

Thus, if we know it is ready, we can check the later stages.